libs/db/badgerdb/db.go

```diff
@@ -0,0 +1,288 @@
+		reverse: opts.Reverse,
+		start:   start,
+		end:     end,
+
+		txn:  txn,
+		iter: iter,
+	}, nil
+}
+
+func (b *BadgerDB) Iterator(start, end []byte) (tmdb.Iterator, error) {
+	opts := badger.DefaultIteratorOptions
+	return b.iteratorOpts(start, end, opts)
+}
+
+func (b *BadgerDB) ReverseIterator(start, end []byte) (tmdb.Iterator, error) {
+	opts := badger.DefaultIteratorOptions
+	opts.Reverse = true
+	return b.iteratorOpts(end, start, opts)
+}
+
+func (b *BadgerDB) Stats() map[string]string {
+	return nil
+}
+
+func (b *BadgerDB) NewBatch() tmdb.Batch {
+	wb := &badgerDBBatch{
+		db:         b.db,
+		wb:         b.db.NewWriteBatch(),
+		firstFlush: make(chan struct{}, 1),
+	}
+	wb.firstFlush <- struct{}{}
+	return wb
+}
+
+var _ tmdb.Batch = (*badgerDBBatch)(nil)
+
+type badgerDBBatch struct {
+	db *badger.DB
+	wb *badger.WriteBatch
+
+	// Calling db.Flush twice panics, so we must keep track of whether we've
+	// flushed already on our own. If Write can receive from the firstFlush
+	// channel, then it's the first and only Flush call we should do.
+	//
+	// Upstream bug report:
+	// https://github.com/dgraph-io/badger/issues/1394
+	firstFlush chan struct{}
+}
+
+func (b *badgerDBBatch) Set(key, value []byte) error {
+	if len(key) == 0 {
+		return tmdb.ErrKeyEmpty
+	}
+	if value == nil {
+		return tmdb.ErrValueNil
+	}
+	return b.wb.Set(key, value)
+}
+
+func (b *badgerDBBatch) Delete(key []byte) error {
+	if len(key) == 0 {
+		return tmdb.ErrKeyEmpty
+	}
+	return b.wb.Delete(key)
+}
+
+func (b *badgerDBBatch) Write() error {
+	select {
+	case <-b.firstFlush:
+		return b.wb.Flush()
+	default:
+		return fmt.Errorf("batch already flushed")
+	}
+}
+
+func (b *badgerDBBatch) WriteSync() error {
+	return withSync(b.db, b.Write())
+}
+
+func (b *badgerDBBatch) Close() error {
+	select {
+	case <-b.firstFlush: // a Flush after Cancel panics too
+	default:
+	}
+	b.wb.Cancel()
+	return nil
+}
+
+type badgerDBIterator struct {
+	reverse    bool
+	start, end []byte
+
+	txn  *badger.Txn
+	iter *badger.Iterator
+
+	lastErr error
+}
+
+func (i *badgerDBIterator) Close() error {
+	i.iter.Close()
+	i.txn.Discard()
+	return nil
+}
+
+func (i *badgerDBIterator) Domain() (start, end []byte) { return i.start, i.end }
+func (i *badgerDBIterator) Error() error                { return i.lastErr }
+
+func (i *badgerDBIterator) Next() {
+	if !i.Valid() {
+		panic("iterator is invalid")
+	}
+	i.iter.Next()
+}
+
+func (i *badgerDBIterator) Valid() bool {
+	if !i.iter.Valid() {
+		return false
+	}
+	if len(i.end) > 0 {
+		key := i.iter.Item().Key()
+		if c := bytes.Compare(key, i.end); (!i.reverse && c >= 0) || (i.reverse && c < 0) {
+			// We're at the end key, or past the end.
+			return false
+		}
+	}
+	return true
+}
+
+func (i *badgerDBIterator) Key() []byte {
+	if !i.Valid() {
+		panic("iterator is invalid")
+	}
+	// Note that we don't use KeyCopy, so this is only valid until the next
+	// call to Next.
+	return i.iter.Item().KeyCopy(nil)
```

Comment:
The comment is confusing 🤔 
> Note that we don't use KeyCopy

but 

> `i.iter.Item().KeyCopy(nil)`

Reply:
This looks to be inorder to comply with the interface DB. If we remove KeyCopy(nil) we return `badger.Item`, this way we return a copy of the key. Nil does an allocation, since we dont know the size of the key, I don't think we can preallocate.